test/unit/acknowledgement/process-invalid.test.js

```diff
@@ -0,0 +1,106 @@
+let mockHoldCategoryName
+let mockHoldCategoryId
+let mockHoldReschedule
+let mockResetPaymentRequestById
+
+let schemeId
+let paymentRequestId
+
+describe('send acknowledgement event', () => {
+  beforeEach(() => {
+    mockConfig.isAlerting = true
+
+    mockAcknowledgement = require('../../mock-acknowledgement')
+    mockAcknowledgementError = require('../../mock-acknowledgement-error')
+    mockHoldCategoryName = 'DAX rejection'
+
+    schemeId = SFI
+    paymentRequestId = 1
+
+    getPaymentRequest.mockReturnValue({
+      schemeId,
+      paymentRequestId,
+      mockFRN: mockAcknowledgementError.frn
+    })
+
+    getHoldCategoryName.mockReturnValue(
+      mockHoldCategoryName
+    )
+
+    getHoldCategoryId.mockReturnValue(
+      mockHoldCategoryId
+    )
+
+    holdAndReschedule.mockReturnValue({
+      mockHoldReschedule
+    })
+
+    resetPaymentRequestById.mockReturnValue({
+      mockResetPaymentRequestById
+    })
+  })
+
+  afterEach(() => {
+    jest.resetAllMocks()
+    jest.resetModules()
+  })
+
+  test('should call sendAcknowledgementEvent when an unsuccessful ack object is given and isAlerting is true', async () => {
+    await processInvalid(schemeId, paymentRequestId, mockFRN, mockAcknowledgementError)
+    expect(mockEvent.sendAcknowledgementEvent).toHaveBeenCalled()
+  })
+
+  test('should call sendAcknowledgementEvent once when an unsuccessful ack object is given and isAlerting is true', async () => {
+    await processInvalid(schemeId, paymentRequestId, mockFRN, mockAcknowledgementError)
+    expect(mockEvent.sendAcknowledgementEvent).toHaveBeenCalledTimes(1)
+  })
+
+  test('should call sendAcknowledgementEvent with a holdCategoryName, an unsuccessful ack object and frn when an unsuccessful ack object is given and isAlerting is true', async () => {
+    await processInvalid(schemeId, paymentRequestId, mockFRN, mockAcknowledgementError)
+    expect(mockEvent.sendAcknowledgementEvent).toHaveBeenCalledWith(mockHoldCategoryName, mockAcknowledgementError, mockFRN)
+  })
+
+  test('should not call sendAcknowledgementEvent when a successful ack object is given and isAlerting is true', async () => {
+    await processInvalid(schemeId, paymentRequestId, mockFRN, mockAcknowledgement)
+    expect(mockEvent.sendAcknowledgementEvent).not.toHaveBeenCalled()
+  })
+
+  test('should not call sendAcknowledgementEvent when an unsuccessful ack object is given but isAlerting is false', async () => {
+    mockConfig.isAlerting = false
+    await processInvalid(schemeId, paymentRequestId, mockFRN, mockAcknowledgementError)
+    expect(mockEvent.sendAcknowledgementEvent).not.toHaveBeenCalled()
+  })
+
+  test('should not call sendAcknowledgementEvent when a successful ack object is given and isAlerting is false', async () => {
+    mockConfig.isAlerting = false
+    await processInvalid(schemeId, paymentRequestId, mockFRN, mockAcknowledgement)
+    expect(mockEvent.sendAcknowledgementEvent).not.toHaveBeenCalled()
+  })
+})
```

Comment:
Missing similar tests for:

- resetPaymentRequestById
- getHoldCategoryName
- getHoldCategoryId
- holdAndReschedule
- transaction.commit
- transaction.rollback
- return of the function itself

Reply:
The code for process-invalid already existed, we only added a bit of code to include our functionality. There wasn't a test file for process-invalid, so I created one and wrote tests for the code I wrote as I figured writing tests for the rest of the code was out of the scope of this ticket and checking in here [...ffc-pay-processing/test-output/lcov-report/app/acknowledgement/index.html] most of the code for process-invalid seemed to already be covered by some other tests (apart from transaction.rollback and throw(error)). Do you think I should still write the tests for the rest of the functions in process-invalid?

Reply:
Question as to why `process-invalid` didn't have a unit tests associated with it is either:
1. doesn't make sense in terms of value (usually because the integration test is enough)
2. whoever implemented it did not include it because... cutting corners/laziness.

You'll have to ask the question and answer it.

If it's 1, fine, let's remove this and add/sign off any integration test cases for this.

However if it's 2, this becomes technical debt. We should/could create a technical debt ticket in the backlog to come back to this. However, we can weigh up the importance/risk, time pressure, size/time required to do it, domain knowledge to understand all possible edge cases (complexity of it) and more to make the decision whether to add it here and now.

Will leave in your judgement.